pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp

```diff
@@ -808,58 +802,124 @@ class ConvertAtenReductionOp : public OpConversionPattern<AtenOpT> {
+    auto selfTy = self.getType().dyn_cast<TensorType>();
 
     if (!selfTy)
-      return op.emitError("Only Tensor types supported in MHLO");
+      return op.emitError("Only tensor types supported in MHLO");
 
     auto outputTy = OpConversionPattern<AtenOpT>::getTypeConverter()
                         ->convertType(op.getType())
-                        .template cast<RankedTensorType>();
+                        .template dyn_cast<RankedTensorType>();
     if (!outputTy)
       return op.emitError(
           "Only ranked tensor type outputs permitted for reduce_mean");
-
-    ElementsAttr reduceDimsAttr;
+    SmallVector<int64_t, 4> reduceDims;
+    DenseIntElementsAttr reduceDimsAttr;
     bool keepDims;
-
     if (failed(readReduceDimsAndKeepDims(
-            op, adaptor, rewriter, reduceDimsAttr, keepDims)))
+            op, adaptor, rewriter, reduceDims, reduceDimsAttr, keepDims)))
       return failure();
 
-    llvm::Optional<Value> result =
-        ConversionFuncT(rewriter, op, outputTy, self, reduceDimsAttr, keepDims);
-
-    if (!result)
+    auto loc = op.getLoc();
+    auto elem_type = selfTy.getElementType();
+    Type type = RankedTensorType::get(/*shape=*/{}, elem_type);
+    DenseElementsAttr const_attr;
+    if (auto float_ty = elem_type.dyn_cast_or_null<mlir::FloatType>()) {
+      mlir::FloatAttr attr = mlir::FloatAttr::get(float_ty, 0);
+      const_attr = mlir::DenseElementsAttr::get(type, attr);
+    } else if (auto int_ty = elem_type.dyn_cast_or_null<mlir::IntegerType>()) {
+      mlir::IntegerAttr attr = mlir::IntegerAttr::get(int_ty, 0);
+      const_attr = mlir::DenseElementsAttr::get(type, attr);
+    } else {
       return failure();
+    }
+
+    auto init_value =
+        rewriter.create<mhlo::ConstOp>(loc, type, const_attr).getResult();
+
+    auto reduction =
+        rewriter.create<mhlo::ReduceOp>(loc, self, init_value, reduceDimsAttr);
 
-    // TBD - support dtype casting.
+    {
+      // Build body
+      OpBuilder::InsertionGuard guard(rewriter);
+      Block* block = rewriter.createBlock(&reduction.body());
 
-    rewriter.replaceOp(op, {result.getValue()});
+      // Block arguments are scalars of the given element type.
+      block->addArguments(type, loc);
+      block->addArguments(type, loc);
+
+      auto reduced = rewriter
+                         .create<MhloOpT>(
+                             loc, block->getArgument(0), block->getArgument(1))
+                         .getResult();
+      rewriter.create<mhlo::ReturnOp>(loc, reduced);
+    }
+    // post proccess for mean & keepDims
+    auto result = reduction.getResult(0);
+    auto dimSizes = mhlo::getDimSizesOfTensor(rewriter, op, self);
+    if (isMean || keepDims) {
+      if (isMean) {
+        auto one = rewriter.create<arith::ConstantOp>(
+            loc, rewriter.getI32IntegerAttr(1));
+        Value numel = one;
+        for (auto& d : dimSizes) {
+          numel = rewriter.create<arith::MulIOp>(loc, numel, d);
```

Comment:
nit: maybe we could have something like `mhlo::getNumElemOfTensor`...

Reply:
done